### ISSUE TITLE
Modules (or circular dependencies) broken in 85?

### ISSUE THREAD

**Issue by drcmda** [mrdoob/three.js#11226]
##### Description of the problem 

Using anything in THREE, even if it's just a Vector

```js
import { Vector3 } from 'three/src/Three';
console.log(Vector3)
```

causes:

    vendor.js:53016 Uncaught TypeError:
        __WEBPACK_IMPORTED_MODULE_0__Vector3__.a is not a constructor

for instance somewhere in Quaternion, which relies on Vector3.

```js
    setFromUnitVectors: function () {

        // http://lolengine.net/blog/2014/02/24/quaternion-from-two-vectors-final

        // assumes direction vectors vFrom and vTo are normalized
>>>>>>> var v1 = new __WEBPACK_IMPORTED_MODULE_0__Vector3__["a" /* Vector3 */]();
```

It looks like for some reason it can't resolve circular dependencies any longer, the internal require returns an empty object. It breaks r85 for module users. The only change i made was updating from 84 to 85.

##### Three.js version

- [x] r85

##### Browser

- [x] Chrome

##### OS

- [x] macOS

##### Build tool

- [x] Webpack 2.4.1

**Comment by Mugen87** [comment]
What happens if your change this line https://github.com/mrdoob/three.js/blob/dev/src/math/Quaternion.js#L354 to just: `var v1;`? There was actually a change in `r85` that creates some helper objects in the forefront.

**Comment by drcmda** [comment]
I think there should be nothing wrong it, it looks correct, it's just that nothing gets resolved for some strange reason.

**Comment by Mugen87** [comment]
The mentioned commit is part of a larger PR, see #10627

That's currently to only source of your error i can image. Let's see what others say.

**Comment by Mugen87** [comment]
Maybe @cecilemuller can help here 😊 ...

**Comment by joe-fr24** [comment]
I had encountered this issue a couple of day ago too, which forced me to roll-back to r84

But I did debug it enough to determine that as @Mugen87 also suggests rolling back the Quaternion v1 Vector3 the instantiation does solve the problem. Depending on transpiling options, both babel and webpack loaders are returning empty objects from __WEBPACK_IMPORTED_MODULE_0__Vector3__["a" /* Vector3 */](); at the point setFromUnitVectors is first called. 

This doesn't really make sense, although I wonder if the change to using Object.assign over prototype definition is also having some kind of affect.

**Comment by drcmda** [comment]
@mrdoob @TristanVALCKE 

@joe-fr24 is it really just the quaternion? It looks like this was part of a huge patch that goes through large parts of the codebase. Wouldn't it just complain about something else if Vector3 is reverted?

And can we be really sure this is webpacks/babels fault since it worked in 84?

**Comment by zorro-fr24** [comment]
@drcmda nope it's not just quaternion:

[revert closure optimisations pr](https://github.com/Flightradar24/three.js/pull/1)

This works for me

**Comment by mrdoob** [comment]
@zorro-fr24 would you like to do a PR here in the meantime?

**Comment by zorro-fr24** [comment]
@mrdoob no problem

**Comment by pailhead** [comment]
I found a circular dependency that broke my jest tests but it seems to be working otherwise. I'm still new to this type of testing/build systems so i'm not sure what the problem is. My case failed with `LineSegments`

**Comment by Mugen87** [comment]
Closing. The original problem was solved with #11246. Other problems in context of circular dependency should be reported in new issues.

**Comment by zorro-fr24** [comment]
FWIW webpack has released version 2.5.1 which resolves the original issue by correctly hoisting exports above imports ( https://github.com/webpack/webpack/issues/4753 ).

This works with r85.2 in my tests

Hopefully this means we won't run into future recurrences.

**Comment by pailhead** [comment]
I think my webpack setup could handle this stuff with no problems, but node couldn't. I'm not super familiar with this stuff, but i've been told that node is harmony compliant, and webpack does more on top of that. Hence webpack could handle the Line import, but jest couldn't.

**Comment by mrdoob** [comment]
> FWIW webpack has released version 2.5.1 which resolves the original issue by correctly hoisting exports above imports ( webpack/webpack#4753 ).

Does this mean that we can now revert #11246?

**Comment by zorro-fr24** [comment]
> Does this mean that we can now revert #11246?

@mrdoob Well I would suggest not - although the very latest revision of webpack 2.x will work now I think, those who still use the 1.x line will probably have issues.

Also consider that Babel has an open issue that looks suspiciously similar. Maybe we can keep an eye on it for r87 ;-)
